MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,361 @@
+        }
+
+        return false
+    }
+
+    func createWalletBackup() throws {
+        guard let backupFolder = TariLib.shared.tariWallet?.publicKey.0?.hex.0 else { throw ICloudBackupWalletError.iCloudContainerNotFound }
+
+        let fileURL = try zipWalletDatabase()
+        let containerURL = try iCloudDirectory().appendingPathComponent(backupFolder)
+
+        if !FileManager.default.fileExists(atPath: containerURL.path) {
+            try FileManager.default.createDirectory(at: containerURL, withIntermediateDirectories: true, attributes: nil)
+        }
+        let backupFileURL = containerURL.appendingPathComponent(fileURL.lastPathComponent)
+        if FileManager.default.fileExists(atPath: backupFileURL.path) {
+            try FileManager.default.removeItem(at: backupFileURL)
+            try FileManager.default.copyItem(at: fileURL, to: backupFileURL)
+        } else {
+            try FileManager.default.copyItem(at: fileURL, to: backupFileURL)
+        }
+        query.operationQueue?.addOperation({ [weak self] in
+            _ = self?.query.start()
+            self?.query.enableUpdates()
+        })
+    }
+
+    func restoreWallet(completion: @escaping (_ error: Error?) -> Void) {
+        let dbDirectory = TariLib.shared.databaseDirectory
+
+        do {
+            let walletFolder = try getLastWalletBackupFolder()
+            try FileManager.default.createDirectory(at: dbDirectory, withIntermediateDirectories: true, attributes: nil)
+
+            restoreBackup(walletFolder: walletFolder, to: dbDirectory) { error in
+                if error != nil {
+                    completion(error)
+                    return
+                }
+
+                let privateKey = PrivateKey()
+                let (privateKeyHex, hexError) = privateKey.hex
+
+                if hexError != nil {
+                    try? FileManager.default.removeItem(at: dbDirectory)
+                    completion(ICloudBackupWalletError.privateKeyError)
+                }
+
+                //Save to keychain and then ensure it's there (If one exists already in the keychain this will overwrite it)
+                TariLib.shared.storedPrivateKey = privateKeyHex
+                guard TariLib.shared.storedPrivateKey != nil else {
+                    TariLogger.error("Failed to save private key to keychain")
+                    try? FileManager.default.removeItem(at: dbDirectory)
+                    completion(ICloudBackupWalletError.privateKeyError)
+                    return
+                }
+                completion(nil)
+            }
+        } catch {
+            completion(error)
+        }
+    }
+}
+
+// MARK: - private methods
+extension ICloudBackup {
+    private func restoreBackup(walletFolder: String, to directory: URL, completion: @escaping ((_ error: Error?) -> Void)) {
+        downloadBackup(walletFolder: walletFolder) { [weak self] (url, error) in
+            guard let zippedBackupURL = url, error == nil else {
```

Comment:
I don't know if it could ever be the case but if `url` and `error` are both `nil`, then this `and` statement will short circuit at `zippedBackupURL` and the next line will call `completion(nil)`.

Reply:
it should never be. But even if it happens we will receive error in completion anyway. Can not see problem here